src/bridges/notional/NotionalBridge.sol

```diff
@@ -0,0 +1,228 @@
+        currencyIds[CETH] = 1;
+        currencyIds[WETH] = 1;
+        currencyIds[DAI] = 2;
+        currencyIds[CDAI] = 2;
+        currencyIds[USDC] = 3;
+        currencyIds[CUSDC] = 3;
+        currencyIds[WBTC] = 4;
+        currencyIds[CWBTC] = 4;
+        underlyingAddrMap[CETH] = WETH;
+        underlyingAddrMap[CDAI] = DAI;
+        underlyingAddrMap[CUSDC] = USDC;
+        underlyingAddrMap[CWBTC] = WBTC;
+        ctokens[CETH] = true;
+        ctokens[CDAI] = true;
+        ctokens[CUSDC] = true;
+        ctokens[CWBTC] = true;
+    }
+
+    receive() external payable {}
+
+    /**
+     * @notice Convert function called by rollup processor, will enter or exit Notional lending position
+     * @dev Will revert if the input asset or the output asset is invalid
+     * @param _inputAssetA The input asset. For entering the market, we support ETH, USDC, DAI, WBTC
+     * and their corresponding ctokens. For exiting the market, we support every valid fcash token.
+     * @param _outputAssetA The output asset. For entering the market, it should be a valid fcash token.
+     For exiting the market, this should be one of ETH, USDC, DAI, WBTC, and their corresponding ctokesn
+     * @param _inputValue The amount of input token
+     * @param _interactionNonce The interaction nonce of the call
+     * @param _auxData For entering the market, this will be the market maturity. For exitng the market,
+     * this will be 0.
+     */
+    function convert(
+        AztecTypes.AztecAsset memory _inputAssetA,
+        AztecTypes.AztecAsset memory,
+        AztecTypes.AztecAsset memory _outputAssetA,
+        AztecTypes.AztecAsset memory,
+        uint256 _inputValue,
+        uint256 _interactionNonce,
+        uint64 _auxData,
+        address
+    )
+        external
+        payable
+        override(BridgeBase)
+        onlyRollup
+        returns (
+            uint256 outputValueA,
+            uint256,
+            bool
+        )
+    {
+        uint40 maturity = uint40(_auxData);
+        _validateInput(maturity, _inputAssetA.erc20Address, _outputAssetA.erc20Address);
+        uint16 currencyId = currencyIds[_inputAssetA.erc20Address];
+        if (currencyId != 0) {
+            if (_inputAssetA.erc20Address == address(0) || _inputAssetA.assetType == AztecTypes.AztecAssetType.ETH) {
+                if (msg.value != _inputValue) {
+                    revert("not enough ether");
+                }
+            }
+            outputValueA = _enter(
+                _inputAssetA.erc20Address,
+                _outputAssetA.erc20Address,
+                currencyId,
+                _inputValue,
+                maturity
+            );
+        } else {
+            outputValueA = _exit(_inputAssetA.erc20Address, _outputAssetA.erc20Address, _inputValue);
+        }
+        if (_outputAssetA.assetType == AztecTypes.AztecAssetType.ERC20) {
+            ERC20(_outputAssetA.erc20Address).approve(ROLLUP_PROCESSOR, outputValueA);
```

Comment:
Since the bridge doesn't hold any tokens it would be safe to pre-approve the tokens. This has the advantage of significantly decreasing gas costs. This would be ideally done in the same function as in which you will load the currencyIds. Check [compound](https://github.com/AztecProtocol/aztec-connect-bridges/blob/a233b8ebd5e8648c100c86de430532bffc3576cf/src/bridges/compound/CompoundBridge.sol#L53) for an inspiration.